### PR TITLE
BIG-bench

### PR DESCRIPTION
**Pull request by andersjohanandreassen**
This PR adds all BIG-bench json tasks to huggingface/datasets.

### CODE REVIEW

**Review by lhoestq — COMMENTED** [review]
Awesome ! Good job with the dataset script, it looks fantastic :)

It looks like the CI is failing on windows because our windows CI is unable to clone the bigbench repository (maybe it has to do with filenames that are longer than 256 characters, which windows don't like). Could the smaller installation of bigbench via pip solve this issue ?
Otherwise we can see how to remove this limitation in our windows CI.

Regarding the dummy data zip files, I think we can just keep `datasets/bigbench/dummy/abstract_narrative_understanding/1.0.0/dummy_data.zip` and remove all the other ones. We just require to have at least one `dummy_data.zip` file.

File: additional-tests-requirements.txt [hunk 1]
```diff
@@ -2,3 +2,15 @@ unbabel-comet>=1.0.0;python_version>'3.6'
 git+https://github.com/google-research/bleurt.git
 git+https://github.com/ns-moosavi/coval.git
 git+https://github.com/hendrycks/math.git
+git+https://github.com/google/BIG-bench.git
```
[the remark this answers is not in the document]
Comment (andersjohanandreassen) [reply additional-tests-requirements.txt:5]: Yes, I will try that! Thanks!

File: datasets/bigbench/bigbench.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    import bigbench.bbseqio.bigbench_bridge as bbb
+    from bigbench.bbseqio import bigbench_json_paths as bb_json_paths
+
+except ModuleNotFoundError as error:
+    logger.info(
+        "Failed to import bigbench. Please see https://github.com/google/BIG-bench or install"
+        "using 'pip install git+https://github.com/google/BIG-bench.git'"
+    )
+    logger.error(error)
+    raise
```
[the remark this answers is not in the document]
Comment (andersjohanandreassen) [reply datasets/bigbench/bigbench.py:38]: Yes, that's much nicer. I will update it asap.

**Comment by andersjohanandreassen** [comment]
> It looks like the CI is failing on windows because our windows CI is unable to clone the bigbench repository (maybe it has to do with filenames that are longer than 256 characters, which windows don't like). Could the smaller installation of bigbench via pip solve this issue ?
> Otherwise we can see how to remove this limitation in our windows CI.
I'm not sure.
If it's git's fault that it can't handle the long filenames, it will possibly be resolved by the pip install. If it's an issue with windows not liking long filenames after it's installed, then it will not be resolved.
I don't have a windows computer to try it on, but I might be able to tweek this PR and do an experiment to find out. 
We're waiting for a quota increase for the pip install (https://github.com/pypa/pypi-support/issues/1782). It's been pending for 2-3 weeks, and I don't have an estimate for when it will be resolved. 

>Regarding the dummy data zip files, I think we can just keep datasets/bigbench/dummy/abstract_narrative_understanding/1.0.0/dummy_data.zip and remove all the other ones. We just require to have at least one dummy_data.zip file.
Sounds great. I will trim that down.

**Comment by lhoestq** [comment]
Do you know what are the other tests dependencies that have conflicts with bigbench ? I can try to split the CI to end up with a compatible list of test dependencies

**Comment by andersjohanandreassen** [comment]
Hi @lhoestq,

I haven't played with eliminating requirements form the test dependencies, and I've been trying to resolve this by modifying the bigbench repo to become compatible. 
In the original bigbench repo, the version requirements were strict, and specifically it had a datasets==1.17.0 requirement which was causing trouble. 
I'm working on PR https://github.com/google/BIG-bench/pull/766 to get some more flexible versions that might be compatible with the test dependencies in  HF/datasets.
We're somewhat flexible in modifying these version numbers if we can figure out what the exact conflict is. 

I've spent some time experimenting with different versions, but I don't have a very efficient way of doing this debugging on my work computer (which for some reason doesn't produce the same sets of errors running python 3.9 instead of 3.6 or 3.7 in the tests). 
It currently fails at 
> The conflict is caused by:
>    bert-score 0.3.6 depends on matplotlib
>    big-bench 0.0.1 depends on matplotlib<4.0 and >=3.5.1
which doesn't seem like it can be the real issue. 

If you have any advice for how to resolve these conflicts, that would be greatly appreciated!

**Comment by andersjohanandreassen** [comment]
Hi again @lhoestq, 
After some more or less random guessing of conflicting packages, I've managed to find a configuration that seems to be compatible with HF/datasets. 

The errors went away after removing version limits on matplotlib and scipy, and loosening numpy from 1.19 -> 1.17 in the bigbench requirements. 

I might do some more tweaking to see if it lets me set some minimal limits on matplotlib and scipy, but I think we at least can move forward.

The WIN tests are still failing, now because of 

> Did not find path entry C:\tools\miniconda3\bin
>C:\tools\miniconda3\envs\py37\python.exe: No module named pytest

I have no way of debugging this locally, and unless there's some way to get more verbose logs, I don't know why it's not finding pytest. Would you be able to take a quick look?

**Comment by andersjohanandreassen** [comment]
One more update on the WIN errors. 
I think all the long filenames are in files in the github repo that does not need to be included. 
We will try to remove them .

**Comment by lhoestq** [comment]
Hi ! The remaining error seems to be a `UnicodeDecodeError` from `setup.py`. I think you can fix your setup.py:
```diff
-     with open(os.path.join(os.path.dirname(__file__), fname)) as f:
+     with open(os.path.join(os.path.dirname(__file__), fname), encoding="uyf-8") as f:

Indeed on windows, when you `open` a file it doesn't always use "utf-8" by default

**Comment by andersjohanandreassen** [comment]
Hi @lhoestq, 
The dependency issues seems to now be resolved 🎉 

Now, the WIN tests are failing at
> ERROR tests/test_arrow_dataset.py::test_dummy_dataset_serialize_s3 - botocore...
> ERROR tests/test_dataset_dict.py::test_dummy_dataset_serialize_s3 - botocore...

Is this testing the dummy dataset that's added in bigbench? If so, I might need some help getting the right format in.

The error message I'm seeing is 
>           raise EndpointConnectionError(endpoint_url=request.url, error=e)
> E           botocore.exceptions.EndpointConnectionError: Could not connect to the endpoint URL: "http://127.0.0.1:5555/test"

Which seems unrelated, but perhaps the real issue is somewhere I'm not seeing?

**Comment by lhoestq** [comment]
Woohoo awesome ! Let me check the CI error

**Comment by lhoestq** [comment]
Can you try to re-run the CI, just in case CircleCI messed up ?

**Comment by andersjohanandreassen** [comment]
Hi @lhoestq, 
Rerunning did not seem to solve the problem. 
It also suddenly triggered an error in the README, but I'm fixing that now. 
The `test_dummy_dataset_serialize_s3` error still seems to remain.

**Comment by andersjohanandreassen** [comment]
Hi again @lhoestq, 
I'm not sure if this is informative or not in terms of debugging, but I deleted the dummy data and the errors for windows still fail and the others still pass. 
Do you have any idea what could be causing this error on windows?

File: datasets/bigbench/README.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+- [Additional Information](#additional-information)
+  - [Dataset Curators](#dataset-curators)
+  - [Licensing Information](#licensing-information)
+  - [Citation Information](#citation-information)
+  - [Contributions](#contributions)
+
+## Dataset Description
+
+- **Homepage/Repository:** [https://github.com/google/BIG-bench](https://github.com/google/BIG-bench)
+- **Paper:** In progress
```
Comment (lhoestq) [inline datasets/bigbench/README.md:75]: Feel free to open a new PR to add the paper once it's ready :)

**Review by lhoestq — APPROVED** [review]
Alright it's all good now !

Thanks a LOT for adding BIG-bench, looking forward to play with it :)

**Comment by lhoestq** [comment]
Now the last question: let's have the dataset under`google/bigbench` @andersjohanandreassen ?

I think it would be nicer, this way you and anyone in your team can update the dataset card whevener you want without going through a github PR. You just need to join the https://huggingface.co/google page :)

**Comment by lhoestq** [comment]
Ok perfect, thank you !

**Comment by lhoestq** [comment]
I noticed that in the latest windows CI run it takes forever to install the dependencies, was there any change in the bigbench dependencies recently ?

**Comment by andersjohanandreassen** [comment]
oh, sorry! I just did a double check on the dependencies, and it seems like there is at least one left that should have been removed.  There's also one new one added. 
Let me get those removed again. Will ping you here when it's updated.

File: .circleci/config.yml [hunk 4]
```diff
@@ -47,6 +47,7 @@ jobs:
                 conda init powershell
                 conda update conda
                 conda create -n py37 python=3.7 pytorch --yes
+                git config --system core.longpaths true
```
Comment (lhoestq) [inline .circleci/config.yml:50]: is this required ? If so, why not have it in run_dataset_script_tests_pyarrow_5_WIN as well ? If not, feel free to remove it

**Review by lhoestq — COMMENTED** [review]
Awesome thanks for fixing the deps :) I think this is ready to merge

File: .circleci/config.yml (hunk 4 above)
Comment (andersjohanandreassen) [reply .circleci/config.yml:50]: It was required for me to install from my github branch during debugging, but it won't be required when installing bigbench from the current tar file (and when we update to the proper pip install). I will remove it.

I really don't understand exactly what is going on with the dependencies, and it seems rather flaky. 
The only thing that's changed between [e356fb6](https://github.com/huggingface/datasets/pull/4125/commits/e356fb66f8255f1752743c2bd5e3ae26a3167515) and the most recent commit [db51cd2](https://github.com/huggingface/datasets/pull/4125/commits/db51cd2df079f15e1e43cf55b184ac3b7655a915) is the addition of explicitly requiring `sacremoses`.
However, if you look back at the tests that passed on commit [890a49f](https://github.com/huggingface/datasets/pull/4125/commits/890a49f52db6c1fe376cd0f3f537d5a8dd4d8430) it installed `sacremoses` without being explicitly listed. 

The bigbench dependencies have remained fixed between [e356fb6](https://github.com/huggingface/datasets/pull/4125/commits/e356fb66f8255f1752743c2bd5e3ae26a3167515) and  [db51cd2](https://github.com/huggingface/datasets/pull/4125/commits/db51cd2df079f15e1e43cf55b184ac3b7655a915), so there must be some other external version update that's suddenly making pip struggle with the windows install.

One observation I made in tracking down the issue is that in the pip install log for the jobs that time out is that it's trying to install the latest `pyarrow` version (version 8.0.0) which is incompatible with the most recent `apache-beam v2.38`. `apache-beam 2.38.0` requires `pyarrow<7.0.0,>=0.15.1`
There is a new `apache-beam v2.39rc2` out as of yesterday which should be made official in the next few days. This is compatible with pyarrow 8.0.0, so maybe that will fix the problem? 
I will try to pin `apache-beam==2.39rc2` as an experiment and remove `sacremoses` to see if this is the issue.

Comment (lhoestq) [reply .circleci/config.yml]: We had to add sacremoses a few days go as well, but it's unrelated to your PR: https://github.com/huggingface/datasets/commit/1a4c185663a6958f48ec69624473fdc154a36a9d

Comment (lhoestq) [reply .circleci/config.yml]: Can you merge upstream/master into your branch and remove the `sacremoses` you added ? We've already added it here:

https://github.com/huggingface/datasets/blob/c1dce72ff259d3e7abee34900e48578ec685fb00/setup.py#L147-L155

The rest looks all good, I think we can merge this PR after this ! Is there anything else you wanted to do before we merge this PR ?

**Comment by lhoestq** [comment]
It looks like there is a circular dependency in `bigbench` at https://storage.googleapis.com/public_research_data/bigbench/bigbench-0.0.1.tar.gz

```python
>>> import bigbench.api.util as bb_utils
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/circleci/.pyenv/versions/3.6.15/lib/python3.6/site-packages/bigbench/api/util.py", line 29, in <module>
    import bigbench.models.query_logging_model as query_logging_model
  File "/home/circleci/.pyenv/versions/3.6.15/lib/python3.6/site-packages/bigbench/models/query_logging_model.py", line 23, in <module>
    import bigbench.api.util as util
AttributeError: module 'bigbench.api' has no attribute 'util'
```

**Review by lhoestq — APPROVED** [review]
Awesome thanks for fixing it ! Let me know if there is something else to do before we merge.

In particular, let me know if you plan to make a release on PyPI so that users can use this version instead of the one hosted on GCS. For now we can merge using the GCS url, but just be careful and don't update it if possible ^^

**Comment by andersjohanandreassen** [comment]
Hi @lhoestq , 
I think we are ready to merge! 

I have one minor question that I haven't been able to figure out: 
Is there a way to bypass the `verify_infos` from triggering? I have `max_examples` as an argument to allow for selecting a fixed subset of the datasets (some of the tasks have *very* many examples). But this is a variable that's not specified by the configs, so it raises an `NonMatchingSplitsSizesError`.
I wasn't able to work my way around this, but perhaps there is a way to bypass this that I'm not seeing?
If this cannot be done, I'm happy to ignore this for now.

Regarding pypi, we are working on a release there, but I'm told there is some issue that there is a problem regarding the upload, and we are not sure when it will be resolved, and it's not in my control. 
I think merging this PR with the GCS is a great idea, and I will open a new PR when the pypi version is ready.

**Comment by lhoestq** [comment]
Cool ! Merging then :D

> Is there a way to bypass the verify_infos from triggering? I have max_examples as an argument to allow for selecting a fixed subset of the datasets (some of the tasks have very many examples). But this is a variable that's not specified by the configs, so it raises an NonMatchingSplitsSizesError.

This is a bug, I opened an issue [here](https://github.com/huggingface/datasets/issues/4462). It should be easy to fix :)

**Pull request merged**